compiler/rustc_attr/src/lib.rs

```diff
@@ -27,6 +27,6 @@ pub use StabilityLevel::*;
 
 pub use rustc_ast::attr::*;
 
-pub(crate) use rustc_ast::HashStableContext;
+pub(crate) use rustc_session::HashStableContext;
```

Comment:
:eyes: :arrow_up: 

I don't fully understand how this works, but without this change, the PR does not compile. Hopefully it is okay.

```console
error[E0277]: the trait bound `__CTX: rustc_session::HashStableContext` is not satisfied
   --> compiler/rustc_attr/src/builtin.rs:151:10
    |
151 |   #[derive(HashStable_Generic)]
    |            ^^^^^^^^^^^^^^^^^^
    |            |
    |            the trait `rustc_session::HashStableContext` is not implemented for `__CTX`
    |            in this derive macro expansion
    |
   ::: .cargo/registry/src/index.crates.io-6f17d22bba15001f/synstructure-0.13.0/src/macros.rs:94:9
    |
94  | /         pub fn $derives(
95  | |             i: $crate::macros::TokenStream
96  | |         ) -> $crate::macros::TokenStream {
    | |________________________________________- in this expansion of `#[derive(HashStable_Generic)]`
    |
    = note: required for `RustcVersion` to implement `HashStable<__CTX>`
```

Reply:
Yeah, this is fine.

Reply:
It's because of the `HashStable_Generic` derive on `RustcVersion` in `rustc_session`. If you *really* wanted to avoid this, you'd need to move that data structure from `rustc_session` to `rustc_ast` or derive it manually.

Deriving it manually may be easier since it's just a few u16s....?

Reply:
Ah yes, this handwritten impl gets the job done: https://github.com/rust-lang/rust/commit/0fc378a6b3c4b214efeab04d3271867c0359542a

```rust
// (in compiler/rustc_session/src/version.rs)

// Handwritten because #[derive(HashStable_Generic)] would generate a stricter
// `Ctx: rustc_session::HashStableContext` bound.
impl<Ctx> HashStable<Ctx> for RustcVersion
where
    Ctx: rustc_ast::HashStableContext,
{
    fn hash_stable(&self, hcx: &mut Ctx, hasher: &mut StableHasher) {
        self.major.hash_stable(hcx, hasher);
        self.minor.hash_stable(hcx, hasher);
        self.patch.hash_stable(hcx, hasher);
    }
}
```

But in the absence of any known downside of changing `rustc_attr::HashStableContext`, I think I would go for keeping the derived impl. I was not able to find any evidence of `dyn HashStableContext` in use anywhere, so I don't see any way that this change could bloat vtables, for example.